Fix item meta mock equality defects

ens-gijs opened the pull request:
# Description
Primary intent is to fix hashcode(), equals(), clone(), copy constructor to act on full ItemMetaMock state.

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Code follows existing style.
- [x] Unit tests added (if applicable).

ens-gijs committed:
Add placeholders for missing methods setDeathScreenScore(int), setDeathScreenScore(int score)

ens-gijs committed:
Removed destroyableKeys and placeableKeys state, which were unused members as they were never exposed via has/get/set methods and has been marked deprecated since 1.20.6

ens-gijs committed:
Fixed bad assumptions of this.enchants nullability in usages, also stumbled into fixing hasConflictingEnchant() impl.

ens-gijs committed:
Synchronize localized name behavior with CraftItemMeta impl. hasLocalizedName() now always returns false and getLocalizedName() delegates to getDisplayName()

ens-gijs committed:
Fixed hashcode(), equals(), clone(), copy constructor to act on full ItemMetaMock state. Reordered operations to match class member def order.
Annotating members for nullability. Some previously nullable members were made non-nullable to simplify hashing logic.

renovate[bot] committed:
fix(deps): update dependency net.bytebuddy:byte-buddy to v1.17.2 (#1277)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Thorinwasher commented:
Tests are failing as, the serialization tests depends on the equals method

ens-gijs commented:
> Tests are failing as, the serialization tests depends on the equals method

And look at that - serialization has been failing to capture and restore all fields too. Fixing.

ens-gijs committed:
Fixed serialize / deserialize on ItemMetaMock.

ens-gijs committed:
Lint - fix curly brace placement

ens-gijs committed:
Lint - fix curly brace placement

4everTheOne commented on src/test/java/org/mockbukkit/mockbukkit/inventory/meta/ItemMetaMockTest.java:
```diff
@@ -1604,6 +1643,39 @@ void givenPossibleValues(ItemRarity value)
+			}
+		}
+
+		@Test
+		void valueIsCopied()
+		{
+			for (ItemRarity rarity : ItemRarity.values())
+			{
+				meta.setRarity(rarity);
+				assertEquals(rarity, new ItemMetaMock(meta).getRarity());
+			}
+		}
```
Please replace this with Parameterized test, so that its easier to identify the faulty rarity.
```java
@ParameterizedTest
void valueIsCopied(ItemRarity rarity)
{
	meta.setRarity(rarity);
	assertEquals(rarity, new ItemMetaMock(meta).getRarity());
}
```

4everTheOne commented on src/test/java/org/mockbukkit/mockbukkit/inventory/meta/ItemMetaMockTest.java:
```diff
@@ -1604,6 +1643,39 @@ void givenPossibleValues(ItemRarity value)
+			}
+		}
+
+		@Test
+		void valueIsCloned()
+		{
+			for (ItemRarity rarity : ItemRarity.values())
+			{
+				meta.setRarity(rarity);
+				assertEquals(rarity, meta.clone().getRarity());
+			}
+		}
```
Please replace this with Parameterized test, so that its easier to identify the faulty rarity.

4everTheOne reviewed (commented):
Just a picky comment, but other than this, it looks good to me

ens-gijs committed:
Parameterize ItemRarity copy/clone tests

ens-gijs committed:
Fix UTF-8 encoding during IDE unit test execution. Fixes failures in ArtMockTest on Windows 10 (for me anyway).

Tip to any future reader: I had to run clean, cleanAllRunTaskCaches, and cleanCustomServiceCaches rules to make this change to into effect.

Thorinwasher committed:
Merge branch 'v1.21' into fix-ItemMetaMock-equality-defects
